pkg/controllers/controlplanemachinesetgenerator/openstack.go

```
@@ -0,0 +1,117 @@
+		return nil, fmt.Errorf("failed to extract failure domains from machines: %w", err)
+	}
+
+	// Fetch failure domains from the machineSets
+	machineSetFailureDomains, err := providerconfig.ExtractFailureDomainsFromMachineSets(machineSets)
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract failure domains from machine sets: %w", err)
+	}
+
+	// We have to get rid of duplicates from the failure domains.
+	// We construct a set from the failure domains, since a set can't have duplicates.
+	failureDomains := failuredomain.NewSet(machineFailureDomains...)
+	// Construction of a union of failure domains of machines and machineSets.
+	failureDomains.Insert(machineSetFailureDomains...)
+
+	openStackFailureDomains := []machinev1.OpenStackFailureDomain{}
+	for _, fd := range failureDomains.List() {
+		openStackFailureDomains = append(openStackFailureDomains, fd.OpenStack())
+	}
+
+	cpmsFailureDomain := machinev1.FailureDomains{
+		OpenStack: &openStackFailureDomains,
+		Platform:  configv1.OpenStackPlatformType,
+	}
+
+	cpmsFailureDomainsApplyConfig := &machinev1builder.FailureDomainsApplyConfiguration{}
+	if err := convertViaJSON(cpmsFailureDomain, cpmsFailureDomainsApplyConfig); err != nil {
+		return nil, fmt.Errorf("failed to convert machinev1.FailureDomains to machinev1builder.FailureDomainsApplyConfiguration: %w", err)
+	}
+
+	return cpmsFailureDomainsApplyConfig, nil
+}
+
+// buildControlPlaneMachineSetOpenStackMachineSpec builds an OpenStack flavored MachineSpec for the ControlPlaneMachineSet.
+func buildControlPlaneMachineSetOpenStackMachineSpec(machines []machinev1beta1.Machine) (*machinev1beta1builder.MachineSpecApplyConfiguration, error) {
+	// The machines slice is sorted by the creation time.
+	// We want to get the provider config for the newest machine.
+	providerConfig, err := providerconfig.NewProviderConfigFromMachineSpec(machines[0].Spec)
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract machine's OpenStack providerSpec: %w", err)
```

Shall we check if all Control plane Machines are equal? As a user, I'd prefer CPMS generation to fail, rather than to override my changes, as unsupported as they may be.